consensus/src/consensus/mod.rs

```diff
@@ -469,12 +469,41 @@ impl Consensus {
     }
 
     fn estimate_network_hashes_per_second_impl(&self, ghostdag_data: &GhostdagData, window_size: usize) -> ConsensusResult<u64> {
-        let window = match self.services.window_manager.block_window(ghostdag_data, WindowType::VaryingWindow(window_size)) {
-            Ok(w) => w,
-            Err(RuleError::InsufficientDaaWindowSize(s)) => return Err(DifficultyError::InsufficientWindowData(s).into()),
-            Err(e) => panic!("unexpected error: {e}"),
-        };
-        Ok(self.services.window_manager.estimate_network_hashes_per_second(window)?)
+        const MIN_WINDOW_SIZE: usize = 1000;
+        if window_size < MIN_WINDOW_SIZE {
+            return Err(ConsensusError::UnderMinWindowSizeAllowed(window_size, MIN_WINDOW_SIZE));
+        }
+
+        let mut count = 0;
+        let mut red_work: BlueWorkType = 0.into();
+        let mut bottom = ghostdag_data.selected_parent;
+        for chain_block in self.services.reachability_service.default_backward_chain_iterator(ghostdag_data.selected_parent) {
+            let gd = self.get_ghostdag_data(chain_block).unwrap();
+            for red in &gd.mergeset_reds {
+                let red_header = self.headers_store.get_header(*red).unwrap();
```

Comment:
Is it safe to assume the header will be in the store for the whole rpc thread execution?

Is there a lock on the store?